fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/ToolbarCfrMiddleware.kt

```diff
@@ -0,0 +1,152 @@
+package org.mozilla.fenix.components.toolbar
+
+import android.view.View
+import androidx.annotation.VisibleForTesting
+import androidx.core.view.isVisible
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.firstOrNull
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.mapNotNull
+import kotlinx.coroutines.flow.transformWhile
+import mozilla.components.browser.state.selector.findCustomTabOrSelectedTab
+import mozilla.components.browser.state.selector.selectedTab
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.browser.toolbar.BrowserToolbar
+import mozilla.components.compose.cfr.CFRPopup
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.lib.state.ext.flowScoped
+import org.mozilla.fenix.R
+import org.mozilla.fenix.shopping.ShoppingExperienceFeature
+import org.mozilla.fenix.utils.Settings
+
+class ToolbarCfrMiddleware(
+    private val browserStore: BrowserStore,
+    private val toolbar: BrowserToolbar,
+    private val settings: Settings,
+    private val shoppingExperienceFeature: ShoppingExperienceFeature,
+    private val customTabId: String?,
+    private val showTcpCfr: () -> Unit,
+    private val showShoppingCFR: (Boolean) -> Unit,
+    private val showEraseCfr: () -> Unit,
+) : Middleware<BrowserFragmentState, BrowserFragmentAction> {
+
+    @VisibleForTesting
+    internal var scope: CoroutineScope? = null
+
+    @VisibleForTesting
+    internal var popup: CFRPopup? = null
+    override fun invoke(
+        context: MiddlewareContext<BrowserFragmentState, BrowserFragmentAction>,
+        next: (BrowserFragmentAction) -> Unit,
+        action: BrowserFragmentAction
+    ) {
+        when (action) {
+            is BrowserFragmentAction.Init -> {
+                val isPrivate = customTabId?.let {
+                    browserStore.state.findCustomTabOrSelectedTab(it)?.content?.private
+                } ?: browserStore.state.selectedTab?.content?.private ?: false
+                val cfr = getCFRToShow(isPrivate)
+                handleCfr(cfr)
+            }
+        }
+    }
```

Comment:
The important piece here is that basically - the view will be created and the middleware will handle checking which CFR should be shown.